src/Network/Wai/Middleware/OpenApi/PathMap.hs

```diff
@@ -0,0 +1,59 @@
+module Network.Wai.Middleware.OpenApi.PathMap
+  ( PathMap
+  , fromOpenApi
+  , lookup
+  ) where
+
+import Prelude hiding (lookup)
+
+import Control.Lens ((^?))
+import Data.Bifunctor (first)
+import Data.ByteString (ByteString)
+import Data.ByteString.Char8 qualified as BS8
+import Data.HashMap.Strict.InsOrd qualified as IHashMap
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.OpenApi (OpenApi, PathItem)
+import Data.OpenApi qualified as OpenApi
+import System.FilePath.Posix qualified as Posix
+
+newtype PathMap = PathMap
+  { unwrap :: Map TemplatedPath PathItem
+  }
```

Comment:
I'd be inclined to give `TemplatedPathComponent` a stock-derived `Eq` and `Ord`, and instead define this as a recursive structure that we do lookups from by walking down the tree one path component at a time.

```suggestion
data PathMap = PathMap
  { root :: Maybe PathItem
  , subpaths :: Map TemplatedPathComponent PathMap
  }
```

Reply:
Hmm, seems neat but I just lifted this from `wai-middleware-validation` and I don't really see anything wrong enough to do such a rewrite.

Patches welcome? If you want, feel free to commit this this PR -- though I do wish I had unit tests on place on this module first to ensure a safe refactor.

Reply:
Hm, maybe will try. I don't *know* that anything is _wrong_, I am just not sure that it is _right_.

Reply:
I have been doing some testing. I'm not sure what the semantics are supposed to be here, but it smells funky.

- If there are patterns `/a` and `/{}`, then a lookup for `/b` will match the wildcard pattern.
- If there are pattern order is flipped  to `/{}` and `/a`, then a lookup for `/b` will return `Nothing`.

Reply:
The doc section [Path Templating Matching](https://swagger.io/specification/#path-templating-matching) attempts to give some rules by illustration.

> Assuming the following paths, the concrete definition, /pets/mine, will be matched first if used:
>
> ```
> /pets/{petId}
> /pets/mine
> ```

This isn't very precisely stated, but it seems to be saying that the path with greater specificity should win, which it more or less what I would have guessed.

It goes on to say

> The following may lead to ambiguous resolution:
> 
> ```
> /{entity}/me
> /books/{id}
> ```

Which seems to just be saying that OpenAPI doesn't care to address the difficult cases and simply suggests "don't make a spec like that."

I don't see anything yet here to suggest that the order in which templates are listed is meant to influence which path matches.

Reply:
Stepping away from the code for a minute, I think what's wrong is actually really simple. When you put `/pets/{petId}` and `/pets/mine` into a `Map`, since `ParameterValue` and `Exact "mine"` compare equal, they're the same map key, so one is going to get overwritten by the other.

Reply:
Yeah, so that being wrong depends on overlapping paths semantics. Since these are ultimately `fromList`ed into the `Map`, the current implementation semantics are "last wins". That's indeed "wrong", since the specification says it should be "most specific wins".

I don't think this is a very difficult bug to fix, or necessarily a reason to abandon the `Map` design: if you gave `TemplatedPathComponent` `<>` such that `Exact` always trumps `ParameterValue`, then `fromListWith (<>)` would prefer more specific paths, making the lookup semantics correct.

Is that "better" than your nested tree idea? Who knows.

Reply:
Well, I went down this hole a bit and it feels quite difficult to do in a Totally Correct way that is also somewhat efficient. I look forward to seeing what you come up with.

I agree the bad `Eq` is compelling, but I think as long as the tests cover all the behaviors, we can be satisfied it works correctly and then decide if it's if it's worth stomaching to get a `Map`'s efficiency (I can't really think of a way to implement fast lookup without relying a `Map`, and so `Eq`, but maybe there is another option).

Reply:
Since the purpose of this package is giving you information to help improve your server, emitting a "you should fix your template overlapping" warning/error on conditions like

- Multiple specs matched this path
- The spec that matched this path was only found after backtracking
 
would probably be a good feature.